[Utils] json helper classes

Twentysix26 opened the pull request:
Since dealing only with DataIO is a pain in the 🍑 I made a couple of helper classes to streamline the process.  
Upon instancing it, it creates the file or loads it if it already exists. Saves automatically at every `.set()`, creates the intermediate folders of the file if requested, a default value can be set in the Server based class...  
I'm still not convinced on the naming and the structure ( @tekulvw proposed to join the two classes ) and I'm most likely gonna extend it, so suggestions are welcome

Twentysix26 committed:
[Utils] Json helper classes

palmtree5 commented:
So I could do something like:

``` python
from blah import blah
import bla
from .utils.json_helper import DataDB, ServerDB

class BlahBlah():
    def __init__(self, bot):
        self.bot = bot
        self.data = ServerDB("data/blahblah/kowlin.json", create_dirs=True)


def setup(bot):
    bot.add_cog(BlahBlah(bot))
```

That's just a very rough example but I can see how this could be useful

calebj committed:
[Utils] Defaultdict, global settings (#495)

Proper defaultdict, added global settings for ServerDB and removed unnecessary re-implementations

Twentysix26 committed:
Empty dict for new servers' entries

AraHaan commented:
There is a such thing as an import hook which can allow you to import a json file with it's file name. This could help as well.

Twentysix26 committed:
Allow `None` as `default_value`

Twentysix26 commented:
Should `.get` allow for a default value? In case nothing is passed, it will either throw an exception or what the `defaultdict` is set as
Edit: That could be a thing for ServerDB's get too

Twentysix26 committed:
Default for .get

Twentysix26 commented on cogs/utils/json_helpers.py:
```diff
@@ -0,0 +1,132 @@
+        self._data[server.id][key] = value
+        self._save()
+
+    def get(self, server, key):
```
Should this have a default? Actually, should *all* `.get_*` functions have a passable default argument?

AraHaan replied:
Yes, it should in case no key is provided.

Also should make this for getting channel data as well if ``channel`` is not None. So I suggest doing something like

```py
    def get(self, server=None, channel=None, key=None):
        """Returns a server or channel's entry"""
        if key is not None:
            if channel is not None:
                if not isinstance(channel, discord.Channel):
                    raise TypeError('Can only get channel or server data')
                return self._data[channel.id][key]
            else:
                if server is not None:
                    if not isinstance(server, discord.Server):
                        raise TypeError('Can only get channel or server data')
                    return self._data[server.id][key]
                else:
                    raise TypeError('Expected a channel or a server object.')
        else:
            raise TypeError('Expected a key, got None instead.')
```

So that way people can tell the bot to explicitly do special things per channel on a server if they want to have it do certain things in certain channels on their server(s).

Twentysix26 commented:
Replaced by config in the rewrite
